pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -557,15 +558,14 @@ func (c *Controller) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.Serv
 			// 2. Headless service without selector
 			out = c.endpoints.GetProxyServiceInstances(c, proxy)
 		} else {
-			var err error
 			// 3. The pod is not present when this is called
 			// due to eventual consistency issues. However, we have a lot of information about the pod from the proxy
 			// metadata already. Because of this, we can still get most of the information we need.
 			// If we cannot accurately construct ServiceInstances from just the metadata, this will return an error and we can
 			// attempt to read the real pod.
 			out, err = c.getProxyServiceInstancesFromMetadata(proxy)
 			if err != nil {
-				log.Warnf("getProxyServiceInstancesFromMetadata for %v failed: %v", proxy.ID, err)
+				err = fmt.Errorf("getProxyServiceInstancesFromMetadata for %v (ip=%v) failed: %v", proxy.ID, proxyIP, err)
```

Comment:
nit: I think this should be a warning. Its not an internal error, its just sometimes not possible from the metadata provided by the client

Reply:
Actually I didn't notice you aren't logging this. We should not return an error here, I think. Was there a reason you think we should?

Reply:
I went back and forth on changing this. I was seeing a lot of these messages in the logs. It didn't seem like an error/warning when using multiple control planes. The check on line 541 will fail because Cluster A's registry won't have any pods for Cluster B. And getProxyServiceInstancesFromMetadata will fail because the remote proxy's cluster_id doesn't match the cluster's ID. It should still be printed but as a debug statement by the aggregate.

Some other alternatives:
* modify getProxyServiceInstancesFromMetadata to not return an error when cluster ID's didn't match. 
* modify the aggregate controller to only call GetProxyServiceInstances() on the controller whose cluster_id matches the target proxy.

Reply:
I am fine with making changes so this logs nothing in multicluster case - that makes complete sense. but if we return an error here, won't the aggregate controller bubble that error up and ultimately we will return an error? So it will always be an error for MC case? I am surprised this works at all, am I missing some step?

Reply:
AFAICT the aggregate controller will log the error but not bubble it up if another controller returned one or more instances (see 
https://github.com/istio/istio/blob/f8464b5eb8f06ee67cf4b1074d33b27610dd9f08/pilot/pkg/serviceregistry/aggregate/controller.go#L266-L270). However, if there is no instances the error eventually bubbles up to initConnection which fails the connection and is not something we want.

How about the following? cc @nmittler 

```diff
diff --git a/pilot/pkg/serviceregistry/aggregate/controller.go b/pilot/pkg/serviceregistry/aggregate/controller.go
index 1741a762e..6964364a4 100644
--- a/pilot/pkg/serviceregistry/aggregate/controller.go
+++ b/pilot/pkg/serviceregistry/aggregate/controller.go
@@ -228,6 +228,9 @@ func (c *Controller) GetProxyServiceInstances(node *model.Proxy) ([]*model.Servi
        // It doesn't make sense for a single proxy to be found in more than one registry.
        // TODO: if otherwise, warning or else what to do about it.
        for _, r := range c.GetRegistries() {
+               if r.Cluster() != "" && node.Metadata != nil && node.Metadata.ClusterID != "" && node.Metadata.ClusterID != r.Cluster() {
+                       continue
+               }
                instances, err := r.GetProxyServiceInstances(node)
                if err != nil {
                        errs = multierror.Append(errs, err)
```

Reply:
Ah thanks... I looked right at that code and missed it. I think that makes sense, probably means things up a bit in general

Reply:
Updated to filter in the aggregate controller. We need to special case the default cluster which is always name `Kubernetes`. I started to clean things up so the default registry uses the provided ClusterName but decided defer that for another PR. This is large enough as-is.